bc_obps/service/facility_service.py

```diff
@@ -16,9 +16,120 @@
+                'longitude_of_largest_emissions',
+            }
+        )
+
+    @classmethod
+    def create_or_update_address(cls, payload: FacilityIn) -> Optional[Address]:
+        """Helper function to create or update an address from payload data."""
+        address_data = payload.dict(
+            include={'street_address', 'municipality', 'province', 'postal_code'}, exclude_none=True
+        )
+        if address_data:
+            return AddressDataAccessService.create_address(address_data)
+        return None
+
+    @classmethod
+    def handle_well_authorization_numbers(cls, user_guid: UUID, payload: FacilityIn, facility: Facility) -> None:
+        """
+        Helper function to process and set well authorization numbers for a facility.
+
+        This method handles the addition and removal of well authorization numbers for a given facility based on
+        the provided payload. It ensures that there are no duplicate well authorization numbers in the new set
+        and updates the facility's well authorization numbers accordingly.
+
+        Args:
+            user_guid (UUID): The GUID of the user making the changes.
+            payload (FacilityIn): The payload containing the new well authorization numbers.
+            facility (Facility): The facility object to be updated.
+
+        Raises:
+            Exception: If there are duplicate well authorization numbers in the new set.
+
+        Steps:
+            1. Extract existing well authorization numbers from the facility.
+            2. Compare new well authorization numbers from the payload with the existing ones.
+            3. Identify and add new well authorization numbers that are not in the existing set.
+            4. Identify and remove old well authorization numbers that are not in the new set.
+        """
+        existing_numbers = set(facility.well_authorization_numbers.values_list('well_authorization_number', flat=True))
+        new_numbers = payload.well_authorization_numbers
+
+        # Check for duplicates within the new_numbers
+        if len(new_numbers) != len(set(new_numbers)):
+            raise Exception(
+                "Well Authorization Number: Well authorization number with this Well authorization number already exists."
+            )
+
+        # Numbers to add
+        numbers_to_add = set(new_numbers) - existing_numbers
+
+        # Add new numbers
+        if numbers_to_add:
+            for number in numbers_to_add:
+                facility.well_authorization_numbers.add(
+                    WellAuthorizationNumberDataAccessService.create_well_authorization_number(user_guid, number)
+                )
+
+        # Numbers to remove
+        numbers_to_remove = existing_numbers - set(new_numbers)
+
+        # Remove old numbers
+        if numbers_to_remove:
+            for number in numbers_to_remove:
+                facility.well_authorization_numbers.filter(well_authorization_number=number).delete()
```

Comment:
Since we are using the `TimeStampedModel` for the `WellAuthorizationNumber` data model, we should archive rather than delete, unless suggested otherwise by @andrea-williams.